### DISCUSSION TITLE
Curl Multi Handle: How to pass data to socket function?

### DISCUSSION THREAD

**Discussion by HallofFamer** [curl/curl#14229] [General]
Hi, I'm trying to implement the functionality similar to the example in [multi-uv](https://curl.se/libcurl/c/multi-uv.html). I dont use global variables so the data such as Curl multi handle, libuv event loop and timer have to be passed as data to callback functions. However, I'm unable to find a way for the socket function `handle_socket(CURL *easy, curl_socket_t s, int action, void *userp, void *socketp)` to accept custom data. Tried something like this: 

```C
curl_context_t* context = create_curl_context();
curl_multi_setopt(curl_handle, CURLMOPT_SOCKETDATA, context);
```

And it does not work, as `socketp` appears to be NULL the first time the call back `handle_socket` is invoked. Why does this happen? Is there an alternative way to pass data to the socket callback function?

**Comment by dfandrich** [comment]
[`CURLMOPT_SOCKETFUNCTION`](https://curl.se/libcurl/c/CURLMOPT_SOCKETFUNCTION.html) is documented to pass the data into the `clientp` argument, which is the equivalent of your `userp` above, not `socketp`. If you want data associated with the socket in `socketp`, you can set it with [`curl_multi_assign`](https://curl.se/libcurl/c/curl_multi_assign.html).

**Comment by HallofFamer** [comment]
@dfandrich 

Thx for letting me know, it seems that this was the issue I was having all along. WIth this being said, I am getting this error when calling `curl_multi_cleanup(curlM)`:

> Assertion failed: !(multi), file D:\Development\lib\multi.c, line 2874

Do you know what could be the cause of the issue? I dont remember ever freeing or cleaning up curl multi handle more than once.
